core/opendaq/synchronization/include/opendaq/sync_component_impl.h

```diff
@@ -0,0 +1,308 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <coretypes/impl.h>
+#include <coretypes/intfs.h>
+#include <coretypes/listobject_factory.h>
+#include <coretypes/listobject_factory.h>
+#include <coretypes/listobject.h>
+#include <coreobjects/property_object.h>
+#include <opendaq/component_impl.h>
+#include <opendaq/sync_component_ptr.h>
+#include <opendaq/sync_component_internal.h>
+
+BEGIN_NAMESPACE_OPENDAQ
+
+template <typename MainInterface, typename ... Interfaces>
+class GenericSyncComponentImpl : public ComponentImpl<ISyncComponentInternal, MainInterface, Interfaces...>
+{
+public:
+    using Super = ComponentImpl<ISyncComponentInternal, MainInterface, Interfaces...>;
+
+    GenericSyncComponentImpl(const ContextPtr& context,
+                             const ComponentPtr& parent,
+                             const StringPtr& localId,
+                             const StringPtr& className = nullptr,
+                             const StringPtr& name = nullptr);
+
+    // ISyncComponent
+    ErrCode INTERFACE_FUNC getSyncLocked(Bool* synchronizationLocked) override;
+    ErrCode INTERFACE_FUNC setSyncLocked(Bool synchronizationLocked) override;
+
+    ErrCode INTERFACE_FUNC getSelectedSource(Int* selectedSource) override;
+    virtual ErrCode INTERFACE_FUNC setSelectedSource(Int selectedSource) override;
+
+    ErrCode INTERFACE_FUNC getInterfaces(IList** interfaces) override;
+    ErrCode INTERFACE_FUNC getInterfaceNames(IList** interfaceNames) override;
+    ErrCode INTERFACE_FUNC addInterface(IPropertyObject* syncInterface) override;
+    ErrCode INTERFACE_FUNC removeInterface(IString* syncInterfaceName) override;
+
+    // ISerializable
+    ErrCode INTERFACE_FUNC getSerializeId(ConstCharPtr* id) const override;
+
+    static ConstCharPtr SerializeId();
+    static ErrCode Deserialize(ISerializedObject* serialized, IBaseObject* context, IFunction* factoryCallback, IBaseObject** obj);
+
+private:
+    template <typename T>
+    typename InterfaceToSmartPtr<T>::SmartPtr getTypedProperty(const StringPtr& name);
+
+};
+
+using SyncComponentImpl = GenericSyncComponentImpl<ISyncComponent>;
+
+
+template <typename MainInterface, typename ... Interfaces>
+GenericSyncComponentImpl<MainInterface, Interfaces...>::GenericSyncComponentImpl(const ContextPtr& context,
+                             const ComponentPtr& parent,
+                             const StringPtr& localId,
+                             const StringPtr& className,
+                             const StringPtr& name)
+    : Super(context, parent, localId, className, name)
+{
+    Super::addProperty(ObjectProperty("Interfaces", PropertyObject()));
+    Super::addProperty(ListProperty("InterfaceNames", List<IString>()));
```

Comment:
This should probably be read-only? It's likely modified by default once an interface is added.

Reply:
we can change the InterfaceNames on add/remove interface

Reply:
Yes, but we do so internally. We can call `setPropertyValueProtected`, while outside users should really not modify this. Honestly, I would even make this property invisible.